### PR TITLE
SE-0055: Making pointer nullability explicit

### PR DESCRIPTION
**Pull request by jrose-apple**
#### What's in this pull request?

An implementation of [SE-0055: Make unsafe pointer nullability explicit using Optional](https://github.com/apple/swift-evolution/blob/master/proposals/0055-optional-unsafe-pointers.md), which touches quite a few things across the compiler and standard library. I'm opening this pull request early even though the review session hasn't completed in anticipation of the proposal being accepted in _some_ form.

---

Before merging this pull request to apple/swift repository:
- [x] The proposal needs to be accepted.
- [x] Tricky bits should be signed off by appropriate reviewers.
- [ ] <del>Test pull request on Swift continuous integration.</del>
- [ ] Have patches ready to go for the other projects (swiftpm, Foundation, etc)

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 1]
```diff
@@ -215,6 +215,26 @@ public func expectEqual<T : Equatable>(expected: T?, _ actual: T?, ${TRACE}) {
   }
 }
 
+public func expectNotEqual<T : Equatable>(
```
Comment (jrose-apple) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: @gribozavr, does this look reasonable?

File: lib/ClangImporter/ImportType.cpp [hunk 2]
```diff
@@ -538,11 +540,8 @@ namespace {
           hint = ImportHint::CFPointer;
         } else if (mappedType->isAnyExistentialType()) { // id, Class
           hint = ImportHint::ObjCPointer;
-        } else if (type->isBlockPointerType()) {
-          // FIXME: This should eventually be "isAnyPointerType", but right now
-          // non-object, non-block pointers are never Optional in Swift; they
-          // just can have a value of 'nil' themselves.
-          hint = ImportHint::CustomNullablePointer;
+        } else if (type->isPointerType() || type->isBlockPointerType()) {
```
Comment (jrose-apple) [inline lib/ClangImporter/ImportType.cpp]: Note for reviewers: the comment was wrong. Clang's `isAnyPointerType` includes pointers-to-ObjC-objects (covered above) and does *not* include block pointers.

**Comment by jrose-apple** [comment]
Review requests:
- @jckarter, @rjmccall, or @slavapestov, can you review the IRGen and Runtime changes?
- @cwillmor, @dgregor, or @jopamer, can you review the Sema changes?
- @moiseev, @dabrahams, or @gribozavr, can you review the stdlib changes? (Sorry…there are many of them and they are relatively boring!)

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 3]
```diff
@@ -661,15 +658,9 @@ final public class NSFastEnumerationIterator : IteratorProtocol {
...
+  static var STACK_BUF_SIZE: Int { return 4 }
 
-  // FIXME: Replace with _CocoaFastEnumerationStackBuf.
-  /// Must have enough space for STACK_BUF_SIZE object references.
-  struct ObjectsBuffer {
-    var buf: (OpaquePointer, OpaquePointer, OpaquePointer, OpaquePointer) =
-      (nil, nil, nil, nil)
-  }
-  var objects: [ObjectsBuffer]
+  var objects: [Unmanaged<AnyObject>?]
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: This code was using inline storage to prevent an extra allocation for the array.  Could you change it back?

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 4]
```diff
@@ -1023,20 +1027,19 @@ extension String {
     scheme tagScheme: String,
     options opts: NSLinguisticTaggerOptions = [],
     orthography: NSOrthography? = nil,
-    tokenRanges: UnsafeMutablePointer<[Range<Index>]> = nil // FIXME:Can this be nil?
+    tokenRanges: UnsafeMutablePointer<[Range<Index>]>? = nil // FIXME:Can this be nil?
   ) -> [String] {
     var nsTokenRanges: NSArray? = nil
-    let result = tokenRanges._withBridgeObject(&nsTokenRanges) {
+    let op = {
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: Why didn't you keep the `_withBridgeObject` API?  You could add it to Optional.

File: stdlib/public/core/ArrayBuffer.swift [hunk 5]
```diff
@@ -261,7 +261,7 @@ extension _ArrayBuffer {
 
       // Tell Cocoa to copy the objects into our storage
       cocoa.buffer.getObjects(
-        UnsafeMutablePointer(result.firstElementAddress),
```
Comment (gribozavr) [inline stdlib/public/core/ArrayBuffer.swift]: Seems like `getObjects` should accept an optional pointer, for the case when it is asked to copy zero objects.

Actually, you could add a fast path to the beginning of this function for that case.  We shouldn't need to do a message send for that.

File: stdlib/public/core/ArrayCast.swift [hunk 6]
```diff
@@ -152,7 +156,7 @@ internal func _arrayConditionalBridgeElements<SourceElement, TargetElement>(
   let buf = _ContiguousArrayBuffer<TargetElement>(
     uninitializedCount: source.count, minimumCapacity: 0)
   
-  var p = buf.firstElementAddress
+  var p = buf.firstElementAddress!
```
Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: `subscriptBaseAddress`?

File: stdlib/public/core/ArrayCast.swift [hunk 7]
```diff
@@ -170,7 +174,7 @@ ElementwiseBridging:
   while false
   
   // Don't destroy anything we never created.
-  buf.count = p - buf.firstElementAddress
+  buf.count = p - buf.firstElementAddress!
```
Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: `subscriptBaseAddress`?

Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: Seems like this is a pattern, would you mind trying it everywhere?

Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: Actually, never mind.  `subscriptBaseAddress` is adjusted for non-zero based indices of slices.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/Foundation.swift]: I'm not sure what you mean. This is still inline storage.

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: Actually, never mind.  `subscriptBaseAddress` is adjusted for non-zero based indices of slices.

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 4 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: I spent a good fifteen minutes trying to figure out what it did, and managed to replace all the similar APIs. I really think just expanding it out inline makes the code easier to read.

File: stdlib/public/core/BridgeObjectiveC.swift [hunk 8]
```diff
@@ -415,26 +408,32 @@ public struct AutoreleasingUnsafeMutablePointer<Pointee /* TODO : class */>
...
   }
 
-  /// Create an instance initialized with `nil`.
-  @_transparent
-  public init(nilLiteral: ()) {
-    self._rawValue = _nilRawPointer
+  /// Explicit construction from an UnsafeMutablePointer.
+  ///
+  /// This is inherently unsafe; UnsafeMutablePointer assumes the
+  /// referenced memory has +1 strong ownership semantics, whereas
```
Comment (gribozavr) [inline stdlib/public/core/BridgeObjectiveC.swift]: UMP does not assume +1.

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (jrose-apple) [inline stdlib/public/core/ArrayCast.swift]: I honestly wasn't sure when the distinction was important to the surrounding code. Do you want me to go back and reanalyze which ones make sense as `firstElementAddress` and which ones make sense as `subscriptBaseAddress`? If the rule is just "we're going to load from it unconditionally", then sure.

File: stdlib/public/core/BridgeObjectiveC.swift (hunk 8 above)
Comment (jrose-apple) [inline stdlib/public/core/BridgeObjectiveC.swift]: It does if it refers to something retain-counted and you try to store to it.

Comment (jrose-apple) [inline stdlib/public/core/BridgeObjectiveC.swift]: Maybe that just falls out of the language rules, but it's worth calling out in *some* form. (I didn't write this text; it's copied from the initializer above.)

File: stdlib/public/core/CocoaArray.swift [hunk 9]
```diff
@@ -47,23 +47,24 @@ internal struct _CocoaArrayWrapper : Collection {
...
+    var dummyBuffer = [AnyObject]()
 
     // This function currently returns nil unless the first
     // subRange.endIndex items are stored contiguously.  This is an
     // acceptable conservative behavior, but could potentially be
     // optimized for other cases.
     let contiguousCount = withUnsafeMutablePointer(&enumerationState) {
-      self.buffer.countByEnumeratingWith($0, objects: nil, count: 0)
+      self.buffer.countByEnumeratingWith(
+        $0, objects: &dummyBuffer, count: 0)
```
Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: We shouldn't need to pass a dummy buffer here.  The function should be annotated as accepting an optional.

File: stdlib/public/core/ContiguousArrayBuffer.swift [hunk 10]
```diff
@@ -256,7 +256,7 @@ public struct _ContiguousArrayBuffer<Element> : _ArrayBufferProtocol {
 
   /// If the elements are stored contiguously, a pointer to the first
   /// element. Otherwise, `nil`.
-  public var firstElementAddress: UnsafeMutablePointer<Element> {
+  public var firstElementAddress: UnsafeMutablePointer<Element>? {
```
Comment (gribozavr) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: Isn't this always non-nil?

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: I am concerned by the array allocation.

Comment (jrose-apple) [inline stdlib/public/core/CocoaArray.swift]: It's not something that's normally invoked by hand, so I don't consider it important to fix. (This is the implementation behind ObjC for-in loops.)

Comment (jrose-apple) [inline stdlib/public/core/CocoaArray.swift]: (It's also part of Foundation, and I didn't want more dependencies.)

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 10 above)
Comment (jrose-apple) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: Unfortunately a non-Optional property will not satisfy an Optional protocol requirement.

File: stdlib/public/core/HashedCollections.swift.gyb [hunk 11]
```diff
@@ -2694,14 +2699,13 @@ final internal class _Native${Self}StorageOwner<${TypeParametersDecl}>
...
-    _ keys: UnsafeMutablePointer<AnyObject>
+    objects: UnsafeMutablePointer<AnyObject>?,
+    _ keys: UnsafeMutablePointer<AnyObject>?
   ) {
     bridgeEverything()
     // The user is expected to provide a buffer of the correct size
     var i = 0 // Position in the input buffer
-    var position = 0 // Position in the dictionary storage
-    let capacity = bridgedNativeStorage.capacity
+    let count = bridgedNativeStorage.capacity
```
Comment (gribozavr) [inline stdlib/public/core/HashedCollections.swift.gyb]: Please keep `capacity` as the variable name.  `count` implies that we have that many entries in the hash table.  In reality, there can be no entries while capacity is non-zero.

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (jrose-apple) [inline stdlib/public/core/CocoaArray.swift]: I would also expect the array allocation to be a mere retain and release, since it's an empty array. But I could pass a dummy address here instead.

File: stdlib/public/core/HashedCollections.swift.gyb (hunk 11 above)
Comment (jrose-apple) [inline stdlib/public/core/HashedCollections.swift.gyb]: Good catch. I think I misread the loops below and thought it actually *was* a count.

File: stdlib/public/core/HashedCollections.swift.gyb [hunk 12]
```diff
@@ -2710,33 +2714,30 @@ final internal class _Native${Self}StorageOwner<${TypeParametersDecl}>
...
         }
       } else {
         // keys nonnull, objects nonnull
-        while position < capacity {
+        for position in 0..<count {
           if bridgedNativeStorage.isInitializedEntry(at: position) {
-            unmanagedObjects[i] = bridgedNativeStorage.value(at: position)
-            unmanagedKeys[i] = bridgedNativeStorage.key(at: position)
+            unmanagedObjects![i] = bridgedNativeStorage.value(at: position)
+            unmanagedKeys![i] = bridgedNativeStorage.key(at: position)
```
Comment (gribozavr) [inline stdlib/public/core/HashedCollections.swift.gyb]: Do you really want to check every time around the loop?  Maybe hoist the check out?

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Or do single-element arrays have some magic I don't know about?

File: stdlib/public/core/Pointer.swift [hunk 13]
```diff
@@ -42,7 +42,33 @@ func _convertInOutToPointerArgument<
...
 }
 
+/// Derive a pointer argument from a value array parameter.
+///
+/// This always produces a non-null pointer, even if the array doesn't have any
+/// storage.
+@_transparent
+@warn_unused_result
+public // COMPILER_INTRINSIC
+func _convertConstArrayToPointerArgument<
```
Comment (gribozavr) [inline stdlib/public/core/Pointer.swift]: Please add unit tests for the new behavior.

File: stdlib/public/core/Runtime.swift.gyb [hunk 14]
```diff
@@ -24,13 +24,17 @@ import SwiftShims
...
-
-  let (oldValue, won) = Builtin.cmpxchg_seqcst_seqcst_RawPointer(
-    target._rawValue, expected.pointee._rawValue, desired._rawValue)
-  expected.pointee._rawValue = oldValue
+  object target: UnsafeMutablePointer<OpaquePointer?>,
+  expected: UnsafeMutablePointer<OpaquePointer?>,
+  desired: OpaquePointer?) -> Bool {
+
+  // Bit-cast directly from 'OpaquePointer?' to 'Builtin.Word' because
+  // Builtin.RawPointer can't be nil.
```
Comment (gribozavr) [inline stdlib/public/core/Runtime.swift.gyb]: SE-0016?

Comment (gribozavr) [inline stdlib/public/core/Runtime.swift.gyb]: You can implement it in underscore land for now.

File: stdlib/public/core/Runtime.swift.gyb [hunk 15]
```diff
@@ -24,13 +24,17 @@ import SwiftShims
...
+  let (oldValue, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+    target._rawValue,
+    unsafeBitCast(expected.pointee, to: Builtin.Word.self),
+    unsafeBitCast(desired, to: Builtin.Word.self))
+  expected.pointee = unsafeBitCast(oldValue, to: Optional<OpaquePointer>.self)
```
Comment (gribozavr) [inline stdlib/public/core/Runtime.swift.gyb]: Why not the `bitPattern` initializer?

File: test/1_stdlib/Runtime.swift [hunk 16]
```diff
@@ -1364,14 +1364,6 @@ Reflection.test("MirrorMirror") {
   expectEqual(0, mirrorMirror.children.count)
 }
 
-Reflection.test("OpaquePointer/null") {
-  // Don't crash on null pointers. rdar://problem/19708338
-  var sequence: OpaquePointer = nil
-  var mirror = Mirror(reflecting: sequence)
-  var child = mirror.children.first!
-  expectEqual("(Opaque Value)", "\(child.1)")
-}
```
Comment (gribozavr) [inline test/1_stdlib/Runtime.swift]: Why not keep the test?

File: stdlib/public/core/UnsafePointer.swift.gyb [hunk 17]
```diff
@@ -74,6 +84,19 @@ public struct ${Self}<Pointee>
...
+  /// Convert from any `UnsafeMutablePointer`, possibly with a
+  /// different `Pointee`.
+  ///
+  /// Returns nil if `from` is nil.
+  ///
+  /// - Warning: the behavior of accesses to pointee as a type
+  ///   different from that to which it was initialized is undefined.
+  @_transparent
+  public init?<U>(_ from : UnsafeMutablePointer<U>?) {
+    guard let unwrapped = from else { return nil }
```
Comment (gribozavr) [inline stdlib/public/core/UnsafePointer.swift.gyb]: Please add tests for new APIs and tests for new behavior in existing initializers.

File: test/ClangModules/cfuncs_parse.swift [hunk 18]
```diff
@@ -73,29 +73,29 @@ func test_pointer() {
   var f: CFloat = 0
   var fa: [CFloat] = [1, 2, 3]
 
-  param_pointer(nil as UnsafeMutablePointer<CInt>)
+  param_pointer(nil as UnsafeMutablePointer<CInt>?)
   param_pointer(&i)
   param_pointer(&ia)
```
Comment (gribozavr) [inline test/ClangModules/cfuncs_parse.swift]: You probably want to add tests for annotated non-nullable pointers.

File: test/Parse/pointer_conversion.swift [hunk 19]
```diff
@@ -5,35 +5,35 @@
...
-func takesAutoreleasingPointer(x: AutoreleasingUnsafeMutablePointer<C>) {} 
+func takesMutablePointer(x: UnsafeMutablePointer<Int>?) {}
+func takesMutableVoidPointer(x: UnsafeMutablePointer<Void>?) {}
+func takesMutableInt8Pointer(x: UnsafeMutablePointer<Int8>?) {} 
+func takesMutableArrayPointer(x: UnsafeMutablePointer<[Int]>?) {} 
+func takesConstPointer(x: UnsafePointer<Int>?) -> Character { return "x" }
+func takesConstInt8Pointer(x: UnsafePointer<Int8>?) {}
+func takesConstUInt8Pointer(x: UnsafePointer<UInt8>?) {}
+func takesConstVoidPointer(x: UnsafePointer<Void>?) {}
+func takesAutoreleasingPointer(x: AutoreleasingUnsafeMutablePointer<C>?) {}
```
Comment (gribozavr) [inline test/Parse/pointer_conversion.swift]: Tests for non-optional pointers?

File: test/PrintAsObjC/classes.swift [hunk 20]
```diff
@@ -253,14 +253,16 @@ typealias AliasForNSRect = NSRect
...
-// CHECK-NEXT: - (void)testBridgingVoid:(void * _Null_unspecified)a b:(void const * _Null_unspecified)b;
+// CHECK-NEXT: - (id _Nonnull * _Nonnull)test:(NSInteger * _Nonnull)a;
+// CHECK-NEXT: - (void)testNested:(NSInteger * _Nonnull * _Nonnull)a;
+// CHECK-NEXT: - (void)testBridging:(NSInteger const * _Nonnull)a b:(NSInteger * _Nonnull)b c:(Methods * _Nonnull * _Nonnull)c;
+// CHECK-NEXT: - (void)testBridgingVoid:(void * _Nonnull)a b:(void const * _Nonnull)b;
 // CHECK-NEXT: init
 // CHECK-NEXT: @end
 @objc class MethodsWithPointers {
-  func test(a: UnsafeMutablePointer<Int>) -> UnsafeMutablePointer<AnyObject> { return nil }
+  func test(a: UnsafeMutablePointer<Int>) -> UnsafeMutablePointer<AnyObject> {
```
Comment (gribozavr) [inline test/PrintAsObjC/classes.swift]: Tests for optional pointers?

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Oh, sorry -- I didn't notice it was an array to begin with.  I thought it used a struct of four pointers directly.  Never mind then.

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: The rule is "we are going to load unconditionally" and "startIndex == 0".  It is the latter part that makes the investigation tricky.

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: It is not a part of Fonudation, it is declared in stdlib/public/core/ShadowProtocols.swift.

Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: And it is not the implementation of ObjC for-in loops, it is `contiguousStorage`, which is called from a subscript in stdlib/public/core/ArrayBuffer.swift.

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 10 above)
Comment (gribozavr) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: Oh, is there a bug for this?

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 4 above)
Comment (dabrahams) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: @jrose-apple Making this instance easier to read is not good enough justification to repeat open-coded logic that is already encapsulated in an idiom we use throughout the library.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/Foundation.swift]: We could probably consider doing that (going back to that?), but I'd make it a separate patch.

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 4 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: I argue that it is a bad idiom, and that two tiny methods, each of which is only used twice, is not enough weight to be worth the encapsulation. Both instances of the related `_withBridgeValue` is easier to understand now; the two instances of `_withBridgeObject` are not quite as clear-cut but still seem like improvements to me. (At the very least my renamed helper function makes a lot more sense.)

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (jrose-apple) [inline stdlib/public/core/ArrayCast.swift]: Right, I didn't want to use `subscriptBaseAddress` in ContiguousArray if the equivalent code in Slice couldn't.

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (jrose-apple) [inline stdlib/public/core/CocoaArray.swift]: The shadow protocols are all supposed to match their original forms unless explicitly specified otherwise, and a mismatch in nullability seems dangerous.

([original method](https://developer.apple.com/library/ios/documentation/Cocoa/Reference/NSFastEnumeration_protocol/index.html#//apple_ref/occ/intfm/NSFastEnumeration/countByEnumeratingWithState:objects:count:))

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 10 above)
Comment (jrose-apple) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: I think it falls under <rdar://problem/17906373> [Protocol conformance] Allow a witness whose type is a subtype of the corresponding requirement's type.

File: test/1_stdlib/Runtime.swift (hunk 16 above)
Comment (jrose-apple) [inline test/1_stdlib/Runtime.swift]: I couldn't figure out a way to write it that wasn't vacuously true or immediately undefined behavior.

File: stdlib/public/core/UnsafePointer.swift.gyb (hunk 17 above)
Comment (jrose-apple) [inline stdlib/public/core/UnsafePointer.swift.gyb]: Yep, next commit. Deliberately not in this commit so that people can ignore test changes if they want.

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: So why can't we explicitly deviate from the original form?  It is not like Foundation can change the behavior here -- there are a lot of Swift programs out there that already pass `nil` here.

I just don't see a reason to create more code for the optimizer to work with, and even possibly introducing a performance regression.

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (gribozavr) [inline stdlib/public/core/ArrayCast.swift]: Why not?  To create more diamonds in the CFG for the optimizer?

File: stdlib/public/core/ContiguousArrayBuffer.swift [hunk 21]
```diff
@@ -323,7 +323,7 @@ public struct _ContiguousArrayBuffer<Element> : _ArrayBufferProtocol {
   @warn_unused_result
   func getElement(i: Int) -> Element {
     _sanityCheck(i >= 0 && i < count, "Array index out of range")
-    return firstElementAddress[i]
+    return firstElementAddress![i]
```
Comment (gribozavr) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: `subscriptBaseAddress[i]`

File: stdlib/public/core/ContiguousArrayBuffer.swift [hunk 22]
```diff
@@ -339,8 +339,8 @@ public struct _ContiguousArrayBuffer<Element> : _ArrayBufferProtocol {
       // firstElementAddress[i] = newValue
       var nv = newValue
       let tmp = nv
-      nv = firstElementAddress[i]
-      firstElementAddress[i] = tmp
+      nv = firstElementAddress![i]
+      firstElementAddress![i] = tmp
```
Comment (gribozavr) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: `subscriptBaseAddress[i]`

File: stdlib/public/core/CocoaArray.swift (hunk 9 above)
Comment (jrose-apple) [inline stdlib/public/core/CocoaArray.swift]: > there are a lot of Swift programs out there that already pass `nil` here

Evidence? Because that's pretty disconcerting. Foundation could certainly change their behavior for new programs. They're unlikely to, but we should not rely on that. "This has always worked in the past" is not a valid reason to violate documented nullability constraints.

Comment (gribozavr) [inline stdlib/public/core/CocoaArray.swift]: > Evidence?

This code in the current and all past versions of the library.

File: test/1_stdlib/Runtime.swift (hunk 16 above)
Comment (gribozavr) [inline test/1_stdlib/Runtime.swift]: `var sequence: OpaquePointer? = nil`

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 10 above)
Comment (jckarter) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: It might be better to define `firstElementAddress` in terms of a non-optional internal property to satisfy the conformance while allowing the implementation to directly use the non-optional property.

File: stdlib/public/core/ArrayCast.swift (hunk 6 above)
Comment (jrose-apple) [inline stdlib/public/core/ArrayCast.swift]: Dmitri, Dave, and I talked about this offline and we decided I should make `firstElementAddress` non-optional, replacing any potentially-nil pointers with the base address of the shared empty array.

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 4 above)
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: The three of us talked about this offline. I'm going to take another look at the `_withBridgeObject` cases again to see if I can simplify the control flow, and I'll put back the helper method (with a better name) if I can't.

File: test/1_stdlib/Runtime.swift (hunk 16 above)
Comment (jrose-apple) [inline test/1_stdlib/Runtime.swift]: Yes, but the mirror reflecting that isn't testing anything new. I guess it's still a likely input, though.

**Comment by jopamer** [comment]
Sema changes look good to me.

**Comment by jrose-apple** [comment]
@swift-ci Please benchmark

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux platform

File: lib/IRGen/GenType.cpp [hunk 23]
```diff
@@ -412,7 +412,52 @@ namespace {
...
+      return true;
+    }
+
+    unsigned getFixedExtraInhabitantCount(IRGenModule &IGM) const override {
+      return 1;
+    }
+
+    APInt getFixedExtraInhabitantValue(IRGenModule &IGM, unsigned bits,
+                                       unsigned index) const override {
+      return APInt(bits, 0);
```
Comment (rjmccall) [inline lib/IRGen/GenType.cpp]: Please assert that index == 0 here.

**Comment by rjmccall** [comment]
Minor adjustment aside, the IRGen parts look good to me.

**Comment by rjmccall** [comment]
Runtime changes LGTM.

**Comment by jrose-apple** [comment]
@swift-ci Please benchmark

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please benchmark

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/SILGen/SILGenFunction.cpp [hunk 24]
```diff
@@ -504,30 +504,31 @@ void SILGenFunction::emitArtificialTopLevel(ClassDecl *mainClass) {
...
+      auto canInnerTy = innerTy->getCanonicalType();
+      unwrappedTy = SILType::getPrimitiveObjectType(canInnerTy);
+    }
+
+    if (unwrappedTy != argv->getType()) {
+      auto converted =
+          emitPointerToPointer(mainClass, ManagedValue::forUnmanaged(argv),
+                               argv->getType().getSwiftRValueType(),
+                               unwrappedTy.getSwiftRValueType());
+      argv = converted.getUnmanagedSingleValue(*this, mainClass);
```
Comment (jrose-apple) [inline lib/SILGen/SILGenFunction.cpp]: @slavapestov, here's the client code for the emitPointerToPointer change. Does this look right?

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

(full testing will fail because it tries to build other repos too)

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test
